test/e2e/consumer_credit_flow_policy.test.ts

```diff
@@ -0,0 +1,170 @@
+    try {
+      sandbox.restore()
+      await client.close()
+      await rabbit.deleteStream(streamName)
+      await rabbit.closeAllConnections()
+      await rabbit.deleteAllQueues({ match: /my-stream-/ })
+    } catch (_e) {}
+  })
+
+  it("NewCreditOnChunkReceived policy requests new credit when chunk is completely received", async () => {
+    const consumerRef = createConsumerRef()
+    const policy = creditsOnChunkReceived(1, 1)
+    sandbox.on(policy, "onChunkReceived")
+    const received: Message[] = []
+    await client.declareConsumer(
+      {
+        stream: streamName,
+        offset: Offset.first(),
+        singleActive: true,
+        consumerRef: consumerRef,
+        creditPolicy: policy,
+      },
+      (message: Message) => {
+        received.push(message)
+      }
+    )
+
+    await send(publisher, chunks)
+
+    await eventually(() => expect(received.length).eql(nMessages))
+    await eventually(() => expect(policy.onChunkReceived).called.exactly(chunks.length))
+    await always(() => expect(policy.onChunkReceived).called.below(chunks.length + 1), 5000)
+  }).timeout(10000)
+
+  it("NewCreditOnChunkProgress policy requests new credit when chunk handled at 50% progress", async () => {
+    const consumerRef = createConsumerRef()
+    const policy = creditsOnChunkProgress(1, 0.5, 1)
+    sandbox.on(policy, "onChunkProgress")
+    const received: Message[] = []
+    await client.declareConsumer(
+      {
+        stream: streamName,
+        offset: Offset.first(),
+        singleActive: true,
+        consumerRef: consumerRef,
+        creditPolicy: policy,
+      },
+      (message: Message) => {
+        received.push(message)
+      }
+    )
+
+    await send(publisher, chunks)
+
+    await eventually(() => expect(received.length).eql(nMessages))
+    await eventually(() => expect(policy.onChunkProgress).called.exactly(nMessages))
+    await always(() => expect(policy.onChunkProgress).called.below(nMessages + 1), 5000)
+  }).timeout(10000)
+
+  it("NewCreditOnChunkProgress policy requests new credit when chunk handled at 100% progress", async () => {
+    const consumerRef = createConsumerRef()
+    const policy = creditsOnChunkProgress(1, 1, 1)
+    sandbox.on(policy, "onChunkProgress")
+    const received: Message[] = []
+    await client.declareConsumer(
+      {
+        stream: streamName,
+        offset: Offset.first(),
+        singleActive: true,
+        consumerRef: consumerRef,
+        creditPolicy: policy,
+      },
+      (message: Message) => {
+        received.push(message)
+      }
+    )
+
+    await send(publisher, chunks)
+
+    await eventually(() => expect(received.length).eql(nMessages))
+    await eventually(() => expect(policy.onChunkProgress).called.exactly(nMessages))
+    await always(() => expect(policy.onChunkProgress).called.below(nMessages + 1), 5000)
+  }).timeout(10000)
+
+  it("NewCreditOnChunkCompleted policy requests new credit when chunk completely handled", async () => {
```

Comment:
Same as before, we can measure that the spy has been called only AFTER all the messages have been handled. If we don't do this, we have a set of actually identical tests which might be problematic.